client/js/src/client.ts

```diff
@@ -821,84 +822,122 @@ export function api_factory(
+												endpoint: _endpoint,
+												fn_index,
+												time: new Date(),
+												...status
+											});
+										} else if (type === "complete") {
+											complete = status;
+										} else if (type == "unexpected_error") {
+											console.error("Unexpected error", status?.message);
+											fire_event({
+												type: "status",
+												stage: "error",
+												message: "An Unexpected Error Occurred!",
+												queue: true,
+												endpoint: _endpoint,
+												fn_index,
+												time: new Date()
+											});
+										} else if (type === "log") {
+											fire_event({
+												type: "log",
+												log: data.log,
+												level: data.level,
+												endpoint: _endpoint,
+												fn_index
+											});
+											return;
+										} else if (type === "generating") {
 											fire_event({
 												type: "status",
 												time: new Date(),
-												...complete,
+												...status,
 												stage: status?.stage!,
 												queue: true,
 												endpoint: _endpoint,
 												fn_index
 											});
 										}
-									}
+										if (data) {
+											fire_event({
+												type: "data",
+												time: new Date(),
+												data: transform_files
+													? transform_output(
+															data.data,
+															api_info,
+															config.root,
+															config.root_url
+													  )
+													: data.data,
+												endpoint: _endpoint,
+												fn_index
+											});
+
+											if (complete) {
+												fire_event({
+													type: "status",
+													time: new Date(),
+													...complete,
+													stage: status?.stage!,
+													queue: true,
+													endpoint: _endpoint,
+													fn_index
+												});
+											}
+										}
 
-									if (status.stage === "complete" || status.stage === "error") {
-										if (event_callbacks[event_id]) {
-											delete event_callbacks[event_id];
-											if (Object.keys(event_callbacks).length === 0) {
-												close_stream();
+										if (
+											status.stage === "complete" ||
+											status.stage === "error"
+										) {
+											if (event_callbacks[event_id]) {
+												delete event_callbacks[event_id];
+												if (Object.keys(event_callbacks).length === 0) {
+													close_stream();
+												}
 											}
 										}
+									} catch (e) {
+										console.error("Unexpected client exception", e);
+										fire_event({
+											type: "status",
+											stage: "error",
+											message: "An Unexpected Error Occurred!",
+											queue: true,
+											endpoint: _endpoint,
+											fn_index,
+											time: new Date()
+										});
+										close_stream();
 									}
 								};
+								event_ids.push(event_id);
 								event_callbacks[event_id] = callback;
+								if (!stream_open) {
```

Comment:
why did you move this?

Reply:
Figured the stream should be opened after the callback got added

Reply:
ok sure, but if a second request is sent while a first one is pending, the stream will be open from the first request, so we shouldn't be expecting it to be the case that the stream is receiving messages for an event only after post request is complete. 
Therefore, it can technically happen that the event stream starts sending messages for an event before the POST request to upload data finishes, and therefore the event callback to handle the message is not ready, which is a race condition that I do not know if we are hitting. I think I do resolve this in a PR I'm working on though